cmd/derper/derper.go

```diff
@@ -513,3 +526,79 @@ func (logFilter) Write(p []byte) (int, error) {
 	log.Printf("%s", p)
 	return len(p), nil
 }
+
+func runDebugServer(addr string, s *derp.Server) {
+	mux := http.NewServeMux()
+	mux.Handle("/debug/vars", http.DefaultServeMux)  // serve out expvar
+	mux.HandleFunc("/debug/varz", tsweb.VarzHandler) // expvar but in prometheus format
+	mux.HandleFunc("/debug/pprof/", pprof.Index)
+	mux.HandleFunc("/debug/pprof/cmdline", pprof.Cmdline)
+	mux.HandleFunc("/debug/pprof/profile", pprof.Profile)
+	mux.HandleFunc("/debug/pprof/symbol", pprof.Symbol)
+	mux.HandleFunc("/debug/pprof/trace", pprof.Trace)
+	mux.Handle("/debug/pprofrate", http.HandlerFunc(handlePprofRate))
+
+	mux.Handle("/", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "text/html; charset=utf-8")
+		w.WriteHeader(200)
+		fmt.Fprintf(w, `<html><body>
+<h1>DERP</h1>
+<p>
+  This is a Tailscale DERP debug server.
+</p>
+<ul>
+  <li>Version: <pre style="display: inline">%s</pre>
+  <li><a href="/debug/vars">/debug/vars</a> (JSON)</li>
+  <li><a href="/debug/varz">/debug/varz</a> (Prometheus)</li>
+  <li><a href="/debug/pprof">/debug/pprof</a></li>
+  <li><a href="/debug/pprof/goroutine?debug=1">/debug/pprof/goroutine?debug=1</a> (goroutine summary)</li>
+</ul>
+`, html.EscapeString(version.Long()))
+	}))
+
+	srv := &http.Server{
+		Addr:    addr,
+		Handler: mux,
+	}
+	if err := srv.ListenAndServe(); err != nil {
+		if err != http.ErrServerClosed {
+			log.Fatal(err)
+		}
+	}
+}
+
+// handlePprofRate adjusts the pprof rate.
+//
+//	$ curl -d block=0 http://localhost:8383/debug/pprofrate
+//	$ curl -d mutex=0 http://localhost:8383/debug/pprofrate
+//
+// See:
+//
+//   - https://pkg.go.dev/runtime#SetMutexProfileFraction
+//   - https://pkg.go.dev/runtime#SetBlockProfileRate
+//
+// And corp#5277.
+func handlePprofRate(w http.ResponseWriter, r *http.Request) {
```

Comment:
This _probably_ needs CSRF protection, though I don't think it's doing anything scary enough to really matter?